### ISSUE TITLE
Error with callr when called from within targets::tar_make() in renv project

### ISSUE THREAD

**Issue by adamaltmejd** [rstudio/renv#773]
See the discussion here: https://github.com/ropensci/targets/discussions/506

and my post here: https://github.com/r-lib/callr/issues/207

Seems most likely this is a renv issue related to what happens when `RENV_CONFIG_USER_PROFILE` is set in `~/.Renviron`.

**Comment by kevinushey** [comment]
Thanks for the bug report -- can you share a reproducible example?

Can you also confirm what the output of:

```
Sys.getenv("RENV_CONFIG_USER_PROFILE")
```

is, within your R session?

**Comment by adamaltmejd** [comment]
The closest I've managed to get to a rep ex is here: https://github.com/ropensci/targets/discussions/506#discussioncomment-847612. @wlandau wasn't able to reproduce though.

```
Sys.getenv("RENV_CONFIG_USER_PROFILE")
[1] "TRUE" 
```

but i get the same error when its set to "FALSE".

**Comment by adamaltmejd** [comment]
Some progress here: https://github.com/ropensci/targets/discussions/506.

Basically there seems to be two distinct issues. One where `tar_make()` fails when there is no newline after `RENV_CONFIG_USER_PROFILE=TRUE` in `~/.Renviron` even though `Sys.getenv("RENV_CONFIG_USER_PROFILE")` shows `"TRUE"`. It seems to be `renv` related since its not just any option in `~/.Renviron` without a newline that creates this but RENV_CONFIG_USER_PROFILE specifically. The error then is:

```
Error in if (!enabled) return(FALSE) : 
  missing value where TRUE/FALSE needed
Calls: source ... renv_bootstrap_load -> <Anonymous> -> renv_load_rprofile
Traceback (most recent calls last):
17: source(".Rprofile", local = TRUE)
16: withVisible(eval(ei, envir))
15: eval(ei, envir)
14: eval(ei, envir)
13: source("renv/activate.R")
12: withVisible(eval(ei, envir))
11: eval(ei, envir)
10: eval(ei, envir)
 9: local(...)
 8: eval.parent(substitute(eval(quote(expr), envir)))
 7: eval(expr, p)
 6: eval(expr, p)
 5: eval(quote(...), new.env())
 4: eval(quote(...), new.env())
 3: renv_bootstrap_load(project, libpath, version)
 2: renv::load(project)
 1: renv_load_rprofile(project)
Execution halted
Error: callr subprocess failed: could not start R, exited with non-zero status, has crashed or was killed
Visit https://books.ropensci.org/targets/debugging.html for debugging advice.
```

Then a separate error when there is a newline that is printed at the end of `tar_make()`. This error only shows if `RENV_CONFIG_USER_PROFILE=TRUE` but the contents of `~/.Rprofile` does not matter. Tested on Mac and Linux.

```
Error in dyn.unload(env$.path) : 
  shared object '/var/folders/b9/b7hn3x0x7pv_1q51k3xmw6m40000gn/T//RtmpimVh4D/callr-client--72f7508.so' was not loaded
Calls: <Anonymous> -> <Anonymous> -> dyn.unload
```

**Comment by kevinushey** [comment]
> Basically there seems to be two distinct issues. One where tar_make() fails when there is no newline after RENV_CONFIG_USER_PROFILE=TRUE in ~/.Renviron even though Sys.getenv("RENV_CONFIG_USER_PROFILE") shows "TRUE".

That is pretty surprising. Is `RENV_CONFIG_USER_PROFILE` the last entry in your `.Renviron`? Does it make a difference if you put it first, or define other variables around it? (Just trying to rule other "weird" possibilities out.)

> Then a separate error when there is a newline that is printed at the end of tar_make(). This error only shows if RENV_CONFIG_USER_PROFILE=TRUE but the contents of ~/.Rprofile does not matter. Tested on Mac and Linux.

This is probably some kind of callr vs. renv incompatibility, but I'd need a reproducible example to be sure (or to know whether the fix lies in renv, callr, or something else)

**Comment by adamaltmejd** [comment]
When I was testing I created an `.Renviron` with only `RENV_CONFIG_USER_PROFILE` set, with and without the newline. The first error only showed up when there was a newline.

Agreed. Let me try to set up a github repo to reproduce it.

**Comment by adamaltmejd** [comment]
Try to clone this and run run.sh. It moves any existing ~/.Rprofile, ~/.Renviron to new files if they exist so you will have to change it back manually. For me, the output of the shell script is:

```
❯ ./run.sh
Error in if (!enabled) return(FALSE) : 
  missing value where TRUE/FALSE needed
Calls: source ... renv_bootstrap_load -> <Anonymous> -> renv_load_rprofile
Traceback (most recent calls last):
17: source(".Rprofile", local = TRUE)
16: withVisible(eval(ei, envir))
15: eval(ei, envir)
14: eval(ei, envir)
13: source("renv/activate.R")
12: withVisible(eval(ei, envir))
11: eval(ei, envir)
10: eval(ei, envir)
 9: local(...)
 8: eval.parent(substitute(eval(quote(expr), envir)))
 7: eval(expr, p)
 6: eval(expr, p)
 5: eval(quote(...), new.env())
 4: eval(quote(...), new.env())
 3: renv_bootstrap_load(project, libpath, version)
 2: renv::load(project)
 1: renv_load_rprofile(project)
Execution halted
Error: callr subprocess failed: could not start R, exited with non-zero status, has crashed or was killed
Visit https://books.ropensci.org/targets/debugging.html for debugging advice.
Execution halted
• start target x
• built target x
• end pipeline
Error in dyn.unload(env$.path) : 
  shared object '/var/folders/b9/b7hn3x0x7pv_1q51k3xmw6m40000gn/T//RtmppSJafG/callr-client--72f7508.so' was not loaded
Calls: <Anonymous> -> <Anonymous> -> dyn.unload
```

**Comment by wlandau** [comment]
Update: I can reproduce both errors from https://github.com/adamaltmejd/renv_callr_targets_error using `callr::r(function() TRUE, show = TRUE)`, so I believe both are unrelated to `targets`. Details: https://github.com/ropensci/targets/discussions/506#discussioncomment-851808.

**Comment by adamaltmejd** [comment]
I'll update the repo to remove the targets dependency to make it easier to debug.

**Comment by petrbouchal** [comment]
I ran into the same issue and can confirm that the RENV_CONFIG_USER_PROFILE=TRUE in ~/.Renviron seems to be the culprit, as is putting `options(renv.config.user.profile = FALSE)` in the project-level .Rprofile (which then overrides RENV_CONFIG_USER_PROFILE=FALSE in ~/.Renviron).

**Comment by adamaltmejd** [comment]
No worries, thanks for the awesome package.

**Comment by kevinushey** [comment]
Okay, I think I know what's going on in the first error. `callr` tries to generate its own `.Renviron` and `.Rprofile` objects to be used by the launched R process. It does so by concatenating the contents of different `.Renviron`s.

If the `.Renviron` is missing a trailing newline, however, then the resulting `.Renviron` is malformed, e.g.

```
RENV_CONFIG_USER_PROFILE
                        TRUER_PROFILE=/var/folders/9n/3nxsgkrj0rlfx196kzdttpww0000gn/T//RtmpjD9Mak/callr-spr-25fa2435a0f1
```

And trying to read that invalid value gives an error. (it would've been helpful if `renv` made that more clear!)

**Comment by kevinushey** [comment]
For the second error, I tried `trace(dyn.unload)` to see if I could figure out what was going on. I see:

```
Tracing function "dyn.load" in package "base"
[1] "dyn.load"
Tracing function "dyn.unload" in package "base"
[1] "dyn.unload"
Tracing dyn.load(file, DLLpath = DLLpath, ...) on entry
[1] "/opt/homebrew/Cellar/r/4.1.2/lib/R/library/utils/libs/utils.so"
Tracing dyn.load(file, DLLpath = DLLpath, ...) on entry
[1] "/opt/homebrew/Cellar/r/4.1.2/lib/R/library/tools/libs/tools.so"
Tracing dyn.load(file, DLLpath = DLLpath, ...) on entry
[1] "/Users/kevin/Library/Caches/org.R-project.R/R/renv/cache/v5/R-4.1/aarch64-apple-darwin20.6.0/processx/3.5.2/0cbca2bc4d16525d009c4dbba156b37c/processx/libs/processx.so"
Tracing dyn.load(file, DLLpath = DLLpath, ...) on entry
[1] "/Users/kevin/Library/Caches/org.R-project.R/R/renv/cache/v5/R-4.1/aarch64-apple-darwin20.6.0/ps/1.6.0/32620e2001c1dce1af49c49dccbb9420/ps/libs/ps.so"
Tracing dyn.load(file, DLLpath = DLLpath, ...) on entry
[1] "/opt/homebrew/Cellar/r/4.1.2/lib/R/library/grDevices/libs/grDevices.so"
Tracing dyn.load(file, DLLpath = DLLpath, ...) on entry
[1] "/opt/homebrew/Cellar/r/4.1.2/lib/R/library/graphics/libs/graphics.so"
Tracing dyn.load(file, DLLpath = DLLpath, ...) on entry
[1] "/opt/homebrew/Cellar/r/4.1.2/lib/R/library/stats/libs/stats.so"
Tracing function "dyn.load" in package "base"
Tracing function "dyn.unload" in package "base"
Tracing dyn.load(file, DLLpath = DLLpath, ...) on entry
[1] "/opt/homebrew/Cellar/r/4.1.2/lib/R/library/tools/libs/tools.so"
Tracing function "dyn.load" in package "base"
Tracing function "dyn.unload" in package "base"
Tracing dyn.load(file, DLLpath = DLLpath, ...) on entry
[1] "/Users/kevin/Library/Caches/org.R-project.R/R/renv/cache/v5/R-4.1/aarch64-apple-darwin20.6.0/processx/3.5.2/0cbca2bc4d16525d009c4dbba156b37c/processx/libs/processx.so"
Tracing dyn.load(file, DLLpath = DLLpath, ...) on entry
[1] "/Users/kevin/Library/Caches/org.R-project.R/R/renv/cache/v5/R-4.1/aarch64-apple-darwin20.6.0/ps/1.6.0/32620e2001c1dce1af49c49dccbb9420/ps/libs/ps.so"
Tracing dyn.load(sofile) on entry
[1] "/var/folders/9n/3nxsgkrj0rlfx196kzdttpww0000gn/T//RtmpD80hWL/callr-client--5873438.so"
Tracing dyn.load(sofile) on entry
[1] "/var/folders/9n/3nxsgkrj0rlfx196kzdttpww0000gn/T//RtmpD80hWL/callr-client--5873438.so"
Tracing dyn.load(file, DLLpath = DLLpath, ...) on entry
[1] "/opt/homebrew/Cellar/r/4.1.2/lib/R/library/grDevices/libs/grDevices.so"
Tracing dyn.unload(env$.path) on entry
[1] "/var/folders/9n/3nxsgkrj0rlfx196kzdttpww0000gn/T//RtmpD80hWL/callr-client--5873438.so"
Tracing dyn.load(file, DLLpath = DLLpath, ...) on entry
[1] "/opt/homebrew/Cellar/r/4.1.2/lib/R/library/graphics/libs/graphics.so"
Tracing dyn.load(file, DLLpath = DLLpath, ...) on entry
[1] "/opt/homebrew/Cellar/r/4.1.2/lib/R/library/stats/libs/stats.so"
Tracing dyn.unload(env$.path) on entry
[1] "/var/folders/9n/3nxsgkrj0rlfx196kzdttpww0000gn/T//RtmpD80hWL/callr-client--5873438.so"
Error in dyn.unload(env$.path) :
  shared object '/var/folders/9n/3nxsgkrj0rlfx196kzdttpww0000gn/T//RtmpD80hWL/callr-client--5873438.so' was not loaded
Calls: <Anonymous> -> <Anonymous> -> dyn.unload
[1] TRUE
```

So it looks like something is triggering the callr client to be loaded twice, and then it's being subsequently unloaded twice. I think the cause is likely that the project `.Rprofile` is being sourced twice, but I'll need to double-check and see why that's happening.

**Comment by kevinushey** [comment]
Okay, I think I know what's happening in the second issue. It's a bad interaction between the way `renv` and `callr` try to manage `.Rprofile`s.

It seems like the following happens:

1. `callr` sets `R_PROFILE` and `R_PROFILE_USER` to point at its own custom `.Rprofile`s.
1. The R session starts in the project directory, and R sources the project `.Rprofile`. This involves activating `renv`.
1. `renv` then tries to load the user profile. Loading that causes `renv` to try sourcing the project `.Rprofile` again.

This is what `callr`'s `.Rprofile` looks like:

```
if (file.exists(".Rprofile")) source(".Rprofile", local = TRUE)
< ... other stuff >
```

Basically, both `renv` and `callr` are trying to source the project `.Rprofile`, and `callr` runs into issues when being loaded twice in this way.

**Comment by kevinushey** [comment]
Should be fixed on main now. Thanks for the report!
